File: experiments/structure/versions.py
#!/usr/bin/env python3
"""How much of a new version of an object is really new. For a pair
(old, new): content-defined chunking (a gear rolling hash, mean chunk
CHUNK bytes) of both; the chunks of `new` already present in `old` are
references, the rest is stored, compressed with zstd -19 (the stand-in
for the level). Also `zstd -19 --patch-from=old new`, which is delta
compression with the old version as the window. Every variant is
rebuilt and compared byte for byte. Sizes: new alone (zstd -19), new
as chunk references + stored chunks, new as a patch, and the share of
bytes new."""
import hashlib, os, subprocess, sys, struct
old_p, new_p = sys.argv[1], sys.argv[2]
CHUNK = int(sys.argv[3]) if len(sys.argv) > 3 else 8192
old = open(old_p, 'rb').read(); new = open(new_p, 'rb').read()
import random
random.seed(1)
GEAR = [random.getrandbits(64) for _ in range(256)]
MASK = (1 << 64) - 1
def chunks(data):
    """Content-defined boundaries: gear hash, cut when the top bits are
    zero (mean CHUNK), min CHUNK/4, max CHUNK*4."""
    bits = CHUNK.bit_length() - 1
    mask = ((1 << bits) - 1) << (64 - bits)
    out = []; start = 0; h = 0; n = len(data); i = 0
    lo, hi = CHUNK // 4, CHUNK * 4
    while i < n:
        h = ((h << 1) + GEAR[data[i]]) & MASK
        i += 1
        if (i - start >= lo and (h & mask) == 0) or i - start >= hi:
            out.append((start, i)); start = i; h = 0
    if start < n: out.append((start, n))
    return out
def z(b):
    return len(subprocess.run(['zstd', '-q', '-19', '-T1', '-c'], input=b, capture_output=True).stdout) if b else 0
oc = chunks(old); nc = chunks(new)
have = {hashlib.sha256(old[a:b]).digest(): (a, b) for a, b in oc}
refs = []; stored = bytearray(); new_bytes = 0
for a, b in nc:
    h = hashlib.sha256(new[a:b]).digest()
    if h in have: refs.append(('old', have[h]))
    else:
        refs.append(('new', (len(stored), len(stored) + b - a))); stored += new[a:b]; new_bytes += b - a
        have[h] = None
# rebuild from refs
out = bytearray()
for kind, (a, b) in refs:
    out += old[a:b] if kind == 'old' else stored[a:b]
assert bytes(out) == new, "rebuild differs"
ref_bytes = len(refs) * 12  # a reference: kind, offset, length as varints/ints, before compression
manifest = b''.join(struct.pack('<BQI', 0 if k == 'old' else 1, a, b - a) for k, (a, b) in refs)
z_new = z(new); z_stored = z(bytes(stored)); z_manifest = z(manifest)
# delta with the old version as the window (zstd --patch-from)
wl = max(20, min(31, (max(len(old), len(new)) - 1).bit_length()))
patch = subprocess.run(['zstd', '-q', '-19', '-T1', f'--long={wl}', '--patch-from=' + old_p, '-c', new_p], capture_output=True).stdout
back = subprocess.run(['zstd', '-q', '-d', f'--long={wl}', '--patch-from=' + old_p, '-c'], input=patch, capture_output=True).stdout
assert back == new, "patch rebuild differs"
name = f"{os.path.basename(old_p)} -> {os.path.basename(new_p)}"
print(f"{name}: new version {len(new):,} B; chunks {len(nc):,} (mean {len(new)//max(1,len(nc)):,} B); bytes not in old version: {100*new_bytes/len(new):.1f}%")
print(f"   new alone, zstd -19:            {z_new:>13,} B  ({len(new)/z_new:.1f}x)")
print(f"   chunk dedup vs old + zstd -19:  {z_stored + z_manifest:>13,} B  ({len(new)/(z_stored+z_manifest):.1f}x; stored chunks {z_stored:,}, manifest {z_manifest:,})  -> {z_new/(z_stored+z_manifest):.1f}x smaller than alone")
print(f"   zstd -19 --patch-from old:      {len(patch):>13,} B  ({len(new)/len(patch):.1f}x)  -> {z_new/len(patch):.1f}x smaller than alone")

File: examples/versions.rs
// How much of a new version of an object is new: content-defined chunk
// dedup against the old version (gear hash, mean CHUNK bytes) with the
// stored chunks zstd -19, and zstd -19 --patch-from (byte-level delta
// with the old version as the window, via the zstd CLI). Byte-exact.
use std::collections::HashMap;
fn chunks(data: &[u8], mean: usize) -> Vec<(usize, usize)> {
    let mut gear = [0u64; 256];
    let mut x = 0x9E37_79B9_7F4A_7C15u64;
    for g in gear.iter_mut() { x ^= x << 13; x ^= x >> 7; x ^= x << 17; *g = x; }
    let bits = mean.trailing_zeros();
    let mask = ((1u64 << bits) - 1) << (64 - bits);
    let (lo, hi) = (mean / 4, mean * 4);
    let mut out = Vec::new();
    let (mut start, mut h) = (0usize, 0u64);
    for i in 0..data.len() {
        h = (h << 1).wrapping_add(gear[data[i] as usize]);
        let len = i + 1 - start;
        if (len >= lo && h & mask == 0) || len >= hi { out.push((start, i + 1)); start = i + 1; h = 0; }
    }
    if start < data.len() { out.push((start, data.len())); }
    out
}
fn sha(b: &[u8]) -> [u8; 16] {
    // A 128-bit content key: two independent 64-bit multiplicative hashes over 8-byte words (a stand-in for SHA-256 in this measurement).
    let (mut a, mut c) = (0x9E37_79B9_7F4A_7C15u64, 0xC2B2_AE3D_27D4_EB4Fu64);
    for w in b.chunks(8) {
        let mut buf = [0u8; 8]; buf[..w.len()].copy_from_slice(w);
        let v = u64::from_le_bytes(buf);
        a = (a ^ v).wrapping_mul(0x9E37_79B9_7F4A_7C15).rotate_left(29);
        c = (c.rotate_left(17) ^ v).wrapping_mul(0xC2B2_AE3D_27D4_EB4F);
    }
    let mut k = [0u8; 16]; k[..8].copy_from_slice(&a.to_le_bytes()); k[8..].copy_from_slice(&c.to_le_bytes()); k
}
fn z19(b: &[u8]) -> usize { if b.is_empty() { 0 } else { zstd::bulk::compress(b, 19).unwrap().len() } }
fn main() {
    let a: Vec<String> = std::env::args().collect();
    let (old_p, new_p) = (&a[1], &a[2]);
    let mean: usize = a.get(3).and_then(|s| s.parse().ok()).unwrap_or(8192);
    let old = std::fs::read(old_p).unwrap(); let new = std::fs::read(new_p).unwrap();
    let oc = chunks(&old, mean); let nc = chunks(&new, mean);
    let mut have: HashMap<[u8; 16], (usize, usize)> = HashMap::with_capacity(oc.len());
    for &(s, e) in &oc { have.entry(sha(&old[s..e])).or_insert((s, e)); }
    let mut stored = Vec::new(); let mut refs: Vec<(bool, usize, usize)> = Vec::new(); let mut new_bytes = 0usize;
    for &(s, e) in &nc {
        let k = sha(&new[s..e]);
        match have.get(&k) {
            Some(&(a, b)) if old[a..b] == new[s..e] => refs.push((true, a, b - a)),
            _ => { refs.push((false, stored.len(), e - s)); stored.extend_from_slice(&new[s..e]); new_bytes += e - s; }
        }
    }
    let mut back = Vec::with_capacity(new.len());
    for &(o, s, l) in &refs { back.extend_from_slice(if o { &old[s..s + l] } else { &stored[s..s + l] }); }
    assert!(back == new, "chunk rebuild differs");
    let mut manifest = Vec::new();
    for &(o, s, l) in &refs { manifest.push(o as u8); manifest.extend_from_slice(&(s as u64).to_le_bytes()); manifest.extend_from_slice(&(l as u32).to_le_bytes()); }
    let (zn, zs, zm) = (z19(&new), z19(&stored), z19(&manifest));
    let wl = ((old.len().max(new.len()) - 1).max(1 << 20)).ilog2() as usize + 1;
    let wl = wl.clamp(20, 31);
    let patch = std::process::Command::new("zstd").args(["-q", "-19", "-T1", &format!("--long={wl}"), &format!("--patch-from={old_p}"), "-c", new_p]).output().unwrap();
    assert!(patch.status.success(), "zstd patch failed: {}", String::from_utf8_lossy(&patch.stderr));
    let patch = patch.stdout;
    let tmp = std::env::temp_dir().join("glyd_patch.zst");
    std::fs::write(&tmp, &patch).unwrap();
    let back = std::process::Command::new("zstd").args(["-q", "-d", &format!("--long={wl}"), &format!("--patch-from={old_p}"), "-c", tmp.to_str().unwrap()]).output().unwrap().stdout;
    assert!(back == new, "patch rebuild differs");
    let name = |p: &str| p.rsplit('/').next().unwrap().to_string();
    println!("{} -> {}: {} B; chunks {} (mean {} B); bytes not in the old version: {:.1}%", name(old_p), name(new_p), new.len(), nc.len(), new.len() / nc.len().max(1), 100.0 * new_bytes as f64 / new.len() as f64);
    println!("   new alone, zstd -19:            {:>14} B ({:.1}x)", zn, new.len() as f64 / zn as f64);
    println!("   chunk dedup vs old + zstd -19:  {:>14} B ({:.1}x)  -> {:.1}x smaller than alone (manifest {} B)", zs + zm, new.len() as f64 / (zs + zm) as f64, zn as f64 / (zs + zm) as f64, zm);
    println!("   zstd -19 --patch-from old:      {:>14} B ({:.1}x)  -> {:.1}x smaller than alone", patch.len(), new.len() as f64 / patch.len() as f64, zn as f64 / patch.len() as f64);
}
